### ISSUE TITLE
tensorflow 2.6 does not detect ptxas.exe

### ISSUE THREAD

**Issue by MJDragoon** [tensorflow/tensorflow#52112]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution: Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary):
- TensorFlow version: 2.6
- Python version: 3.9
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source): 
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.4
- GPU model and memory: GTX 2060 6GB

**Describe the problem**

I'm currently having problems with Tensorflow not detecting the ptxas.exe inside the cuda folders as seen here:

`2021-09-23 23:36:12.332220: I tensorflow/stream_executor/cuda/cuda_dnn.cc:369] Loaded cuDNN version 8204
2021-09-23 23:36:12.941230: E tensorflow/core/platform/windows/subprocess.cc:287] Call to CreateProcess failed. Error code: 2
2021-09-23 23:36:12.944774: E tensorflow/core/platform/windows/subprocess.cc:287] Call to CreateProcess failed. Error code: 2
2021-09-23 23:36:12.947274: W tensorflow/stream_executor/gpu/asm_compiler.cc:77] Couldn't get ptxas version string: Internal: Couldn't invoke ptxas.exe --version
2021-09-23 23:36:12.951640: E tensorflow/core/platform/windows/subprocess.cc:287] Call to CreateProcess failed. Error code: 2
2021-09-23 23:36:12.953523: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] Internal: Failed to launch ptxas
Relying on driver to perform ptx compilation.
Modify $PATH to customize ptxas location.`

But I can confirm that the ptxas.exe does exist.

I'm currently new to this kind of stuff, so if possible elaborate more so I could understand better, thank you.

**Provide the exact sequence of commands / steps that you executed before running into the problem**

``import os
from pickle import TRUE
os.add_dll_directory("C:\\Program Files\\NVIDIA GPU Computing Toolkit\\CUDA\\v11.4\\bin")
import keras
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
import numpy as np
import tensorflow as tf

batch_size = 128
num_classes = 10
epochs = 12

img_rows, img_cols = 28, 28

(x_train, y_train), (x_test, y_test) = mnist.load_data()

x_train = x_train.reshape(60000,28,28,1)
x_test = x_test.reshape(10000,28,28,1)

print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=(28,28,1)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=tf.keras.optimizers.Adadelta(),
              metrics=['accuracy'])

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])``

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

Here is the complete output:
`x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
2021-09-23 23:36:10.447932: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-09-23 23:36:10.895610: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1510] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 3967 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2060, pci bus id: 0000:01:00.0, compute capability: 7.52021-09-23 23:36:11.489872: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
Epoch 1/12
2021-09-23 23:36:12.332220: I tensorflow/stream_executor/cuda/cuda_dnn.cc:369] Loaded cuDNN version 8204
2021-09-23 23:36:12.941230: E tensorflow/core/platform/windows/subprocess.cc:287] Call to CreateProcess failed. Error code: 2
2021-09-23 23:36:12.944774: E tensorflow/core/platform/windows/subprocess.cc:287] Call to CreateProcess failed. Error code: 2
2021-09-23 23:36:12.947274: W tensorflow/stream_executor/gpu/asm_compiler.cc:77] Couldn't get ptxas version string: Internal: Couldn't invoke ptxas.exe --version
2021-09-23 23:36:12.951640: E tensorflow/core/platform/windows/subprocess.cc:287] Call to CreateProcess failed. Error code: 2
2021-09-23 23:36:12.953523: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] Internal: Failed to launch ptxas
Relying on driver to perform ptx compilation.
Modify $PATH to customize ptxas location.
This message will be only logged once.
469/469 [==============================] - 7s 9ms/step - loss: 29.2319 - accuracy: 0.1548 - val_loss: 3.7147 - val_accuracy: 0.4812
Epoch 2/12
469/469 [==============================] - 4s 9ms/step - loss: 12.0754 - accuracy: 0.2700 - val_loss: 1.5613 - val_accuracy: 0.6339
Epoch 3/12
469/469 [==============================] - 4s 9ms/step - loss: 6.1234 - accuracy: 0.3334 - val_loss: 1.2056 - val_accuracy: 0.6057
Epoch 4/12
469/469 [==============================] - 4s 9ms/step - loss: 3.6699 - accuracy: 0.3449 - val_loss: 1.4396 - val_accuracy: 0.5644
Epoch 5/12
Epoch 6/12
469/469 [==============================] - 4s 9ms/step - loss: 2.2604 - accuracy: 0.3577 - val_loss: 1.6565 - val_accuracy: 0.5153
Epoch 7/12
469/469 [==============================] - 4s 9ms/step - loss: 2.0842 - accuracy: 0.3743 - val_loss: 1.6186 - val_accuracy: 0.5310
Epoch 8/12
469/469 [==============================] - 4s 9ms/step - loss: 1.9903 - accuracy: 0.3897 - val_loss: 1.5678 - val_accuracy: 0.5573
Epoch 9/12
469/469 [==============================] - 4s 9ms/step - loss: 1.9153 - accuracy: 0.4106 - val_loss: 1.4787 - val_accuracy: 0.5954
Epoch 10/12
469/469 [==============================] - 4s 9ms/step - loss: 1.8446 - accuracy: 0.4291 - val_loss: 1.4024 - val_accuracy: 0.6281
Epoch 11/12
469/469 [==============================] - 4s 9ms/step - loss: 1.7916 - accuracy: 0.4465 - val_loss: 1.3342 - val_accuracy: 0.6566
Epoch 12/12
469/469 [==============================] - 4s 10ms/step - loss: 1.7438 - accuracy: 0.4593 - val_loss: 1.2915 - val_accuracy: 0.6798
Test loss: 1.2914707660675049
Test accuracy: 0.6797999739646912`

**Comment by mohantym** [comment]
Hi @MJDragoon , The above code ran successfully after removing this line . 
`os.add_dll_directory("C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4\bin")`

 It seems you are trying to add path for CUDA in code itself. 
Can you try again after setting it as environment path from environment variables instead. 

Attaching [GIST ](https://colab.research.google.com/gist/mohantym/e5f00a636a87792d0f878975ed46bbb4/github_52122.ipynb) in 2.6 for reference

**Comment by MJDragoon** [comment]
> Hi @MJDragoon , The above code ran successfully after removing this line .
> `os.add_dll_directory("C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4\bin")`
> 
> It seems you are trying to add path for CUDA in code itself.
> Can you try again after setting it as environment path from environment variables instead.
> 
> Attaching [GIST ](https://colab.research.google.com/gist/mohantym/e5f00a636a87792d0f878975ed46bbb4/github_52122.ipynb) in 2.6 for reference

Hi @mohantym, thanks for the reply but unfortunately, if I removed the line you suggested I would get these errors

`2021-09-24 08:57:06.355128: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2021-09-24 08:57:06.357962: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cublas64_11.dll'; dlerror: cublas64_11.dll not found
2021-09-24 08:57:06.361187: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cublasLt64_11.dll'; dlerror: cublasLt64_11.dll not found
2021-09-24 08:57:06.363529: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cufft64_10.dll'; dlerror: cufft64_10.dll not found
2021-09-24 08:57:06.366463: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'curand64_10.dll'; dlerror: curand64_10.dll not found
2021-09-24 08:57:06.368716: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cusolver64_11.dll'; dlerror: cusolver64_11.dll not found
2021-09-24 08:57:06.371322: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cusparse64_11.dll'; dlerror: cusparse64_11.dll not found
2021-09-24 08:57:06.375071: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudnn64_8.dll'; dlerror: cudnn64_8.dll not found`

**Comment by mohantym** [comment]
Hi @MJDragoon ,Can you look at these threads for answer. [link1](https://github.com/tensorflow/tensorflow/issues/46606#issuecomment-765516590), [setting environment variable for cuda](https://docs.nvidia.com/gameworks/content/developertools/desktop/environment_variables.htm). Thanks!

**Comment by MJDragoon** [comment]
Hi @mohantym, thank you for the reply.

I have checked the links you posted and was sure I added the path in the environment variables when I installed it, nonetheless, I still went through and did the solutions in those links but unfortunately it did not fix the problem. I still needed the os.add_dll_directory in order to access the dll's it didn't find, and it did not detect the ptxas.exe within CUDA.

**Comment by mohantym** [comment]
Hi @MJDragoon  ,Could you please try again with CUDA 11.2 , cuDNN8.1 and Python 3.8 ?
Reference:
https://www.tensorflow.org/install/gpu

**Comment by MJDragoon** [comment]
Hi @mohantym.

Thankfully that did the trick! but I still needed the os.add_dll_directory otherwise errors will appear, other than that no errors seem to occur! thank you for your time.

**Comment by mohantym** [comment]
Ok @MJDragoon,  If the above steps worked , Can you try again after adding  **os.add_dll_directory()** operation 
to see if error occurs ?

**Comment by MJDragoon** [comment]
Hi @mohantym,

After adding the os.add_dll_directory() into the code, no ptxas.exe error occurred, but I did get a cudart64_110.dll error. I am currently reinstalling everything that I have downloaded during this project in order to at least fix that issue.

**Comment by mohantym** [comment]
Hi @sanatmpa1! 
Could you please look at this issue! Issue is not replicating in Colab ,when commenting out this line .Attaching GIST for reference in [2.5](https://colab.research.google.com/gist/mohantym/ab6a4daf4454ec6ee6c42880fcd49767/52112.ipynb#scrollTo=KXnXiymmrI7U),[2.6](https://colab.research.google.com/gist/mohantym/166ce616e3aa6899c443da0f4939d193/52112.ipynb#scrollTo=KXnXiymmrI7U) and [2.7.0dev(nightly)](https://colab.research.google.com/gist/mohantym/fbf32de1df79ec8415f1788a2d90a996/52112.ipynb#scrollTo=KXnXiymmrI7U)

>  os.add_dll_directory("C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.4\bin")

**Comment by sanatmpa1** [comment]
@MJDragoon,

For the error related to cudart64_110.dll , you can refer to the similar [issue1](cudart64_110.dll ), [issue2](https://github.com/tensorflow/tensorflow/issues/43193). As you said you can try to do a fresh install and make sure you have the tested build configurations as per this [guide](https://www.tensorflow.org/install/source_windows#gpu) while installing. Thanks!
